### PR TITLE
Local scale writers for partitioned data

### PR DESCRIPTION
**Pull request by gaurav8297**
The approach and problem are documented here: https://github.com/trinodb/trino/issues/13379

Note: Few tests are still pending which I'm working on

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine

> How would you describe this change to a non-technical end user or system administrator?

Increase the performance of partitioned writes with skewness.

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ScaleWriterPartitioningExchanger
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:32]: It's not clear to me what this class does. Could add a comment describing the intended behaviour?

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                continue;
+            }
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            if (bucketSize == page.getPositionCount()) {
+                // entire page will be sent to this partition, compact and send the page after releasing the lock
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:139]: is there any lock help here?

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+            if (pageSink == null) {
+                pageSink = pageSinkSupplier.get();
+                partitionPageSinks[partition] = pageSink;
+            }
+
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            if (partitionSize == page.getPositionCount()) {
+                // entire page will be sent to this partition, compact and send the page after releasing the lock
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:375]: is this comment intended to be here? there is no compact, nor lock...

**Review by lukasz-stec — COMMENTED** [review]
some initial comments

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java (hunk 2 above)
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java]: No lock is required.

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java (hunk 1 above)
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:58]: Added comments to this class. Also, Please take a look at the approach in this issue: https://github.com/trinodb/trino/issues/13379

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+        timer.end(statisticsTiming);
+        ListenableFuture<Void> blockedOnAggregation = statisticAggregationOperator.isBlocked();
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedOnWrites = ImmutableList.builder();
+        blockedOnWrites.add(blockedOnAggregation);
+
+        Page partitionedPage = partitionPagePreparer.apply(page);
+
+        for (int position = 0; position < partitionedPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionedPage, position);
```
[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:355]: Partitioning is required here because it is possible that `ScaleWriterPartitioningExchanger` sends out pages which contain multiple partitions. Basically, `ScaleWriterPartitioningExchanger` assign multiple set of partitions to a single writer after considering scaling stuff, and then create a page which contains all these partitions.

Another approach could be to create a page per partition which include the partition unique id as a separate RLE channel. However, it increases the complexity since you need to describe this channel as part of plan and there can be very small pages which can impact performance. Therefore we've decided not to go with this soln.

In future, we could pre-generate the hashes to save the computation. https://github.com/trinodb/trino/issues/14184

[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:355]: I'll add more comments to this class

**Comment by gaurav8297** [comment]
**Benchmarks**

session:
preferred_write_partitioning_min_number_of_partitions=1
task_writer_count=2
task_scale_writers_min_writer_count=2
task_scale_writers_max_writer_count=8
task_scale_writers_partition_count=128

`lineitem_test` is partitioned over `shipmode` column which contains 7 big partitions.

Before (scaling disabled):
```
trino:insert_demo> Insert into lineitem_test select orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, commitdate, rec
eiptdate, shipinstruct, shipmode FROM tpch_sf100_orc_part.lineitem;
INSERT: 600037902 rows

Query 20220928_230049_00011_6j66t, FINISHED, 7 nodes
Splits: 2,763 total, 2,763 done (100.00%)
6:00 [600M rows, 15.6GB] [1.66M rows/s, 44.5MB/s]
```

After (scaling enabled):
```
trino:insert_demo> Insert into lineitem_test select orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, commitdate, rec
eiptdate, shipinstruct, shipmode FROM tpch_sf100_orc_part.lineitem;
INSERT: 600037902 rows

Query 20220928_225723_00008_6j66t, FINISHED, 7 nodes
Splits: 2,799 total, 2,799 done (100.00%)
2:45 [600M rows, 15.6GB] [3.63M rows/s, 96.9MB/s]
```

More than 2x improvements

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java (hunk 1 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:58]: Nice description!

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:355]: > it is possible that ScaleWriterPartitioningExchanger sends out pages which contain multiple partitions

Ok, got it. 
But when I look at `HivePageSink` or `IcebergPageSink` I see it partitions the page anyway because it has to split the data into files so why partitioning here + multiple `ConnectorPageSink`s is needed? Should we noty leave it to the `ConnectorPageSink`?

[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:355]: We need partitioning here along with multiple `ConnectorPageSinks` to collect statistics around partition level physicaWrittenBytes which is essential for scaling. In the ideal world, we could get rid of partitioning from pageSink if we do partitioning directly inside the local exchange itself using pageIndexer which handles hash collisions also. However, it again adds complexity which I mentioned in the above comment.

File: core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    private final NodePartitioningManager nodePartitioningManager;
+    private final BlockTypeOperators blockTypeOperators;
+
+    public PartitionFunctionFactory(NodePartitioningManager nodePartitioningManager, BlockTypeOperators blockTypeOperators)
+    {
+        this.nodePartitioningManager = requireNonNull(nodePartitioningManager, "nodePartitioningManager is null");
+        this.blockTypeOperators = requireNonNull(blockTypeOperators, "blockTypeOperators is null");
+    }
+
+    public PartitionFunction create(
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java:45]: this is just refactor, right?

File: core/trino-main/src/main/java/io/trino/operator/DriverContext.java [hunk 6]
```diff
@@ -275,6 +276,25 @@ public long getPhysicalWrittenDataSize()
         return physicalWrittenBytes;
     }
 
+    public long[] getPartitionPhysicalWrittenDataSize()
+    {
+        // Avoid using stream api for performance reasons
+        int maxSize = 0;
+        List<long[]> partitionPhysicalWrittenDataSizes = new ArrayList<>();
+        for (OperatorContext context : operatorContexts) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/DriverContext.java:284]: there should only be one operator in driver that produces `getPartitionPhysicalWrittenDataSize()`,
so appending code seems to be noop

File: core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java (hunk 5 above)
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java:45]: This is required because we are using this in both `LocalExchange` as well as `PartitionTableWriterOperator`

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 7]
```diff
@@ -80,6 +80,8 @@
     // and acceptable in terms of resource utilization since values like 32 or higher could take
     // more resources, hence potentially affect the other concurrent queries in the cluster.
     private int scaleWritersMaxWriterCount = 8;
+    private int scaleWritersMinWriterCount = 2;
+    private int scaleWritersPartitionCount = 128;
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:84]: How this value is related to `scaleWritersMaxWriterCount`?

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.columnChannels = requireNonNull(columnChannels, "columnChannels is null");
+            this.pageSinkManager = requireNonNull(pageSinkManager, "pageSinkManager is null");
+            checkArgument(
+                    writerTarget instanceof CreateTarget
+                            || writerTarget instanceof InsertTarget
+                            || writerTarget instanceof RefreshMaterializedViewTarget
+                            || writerTarget instanceof TableExecuteTarget,
+                    "writerTarget must be CreateTarget, InsertTarget, RefreshMaterializedViewTarget or TableExecuteTarget");
+            this.target = requireNonNull(writerTarget, "writerTarget is null");
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:125]: `requireNonNull` is needed here? we tested the object with `instanceof`.

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+import static io.trino.sql.planner.plan.TableWriterNode.WriterTarget;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.function.Function.identity;
+
+/**
+ * It partitions the pages and then create a separate pageSink for each partitioned page.
+ * Therefore, it allows us to track physicalWrittenBytes per partition.
+ */
+public class PartitionedTableWriterOperator
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:79]: There is a lot of overlap in functionality and a lot of code copied from `TableWriterOperator`. Can this be somehow reused? This could improve the readability of this class by removing non-essential functionality.

File: core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java [hunk 10]
```diff
@@ -135,18 +130,52 @@ else if (partitioning.equals(SCALED_WRITER_DISTRIBUTION)) {
...
+
+                PartitionFunction partitionFunction = partitionFunctionFactory.create(
+                        session,
+                        unwrappedPartitioning,
+                        partitionChannels,
+                        partitionChannelTypes,
+                        partitionHashChannel,
+                        bucketCount);
+                Function<Page, Page> partitionPagePreparer;
+                if (isSystemPartitioning(unwrappedPartitioning)) {
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java:150]: Consider extracting this to a method (e.g. `createPartitionPagePreparer`) and reusing it in the following `if`

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Consumer<PageReference> wholePagePartition = partitionPage(page, partitionedPagePreparer.apply(page));
+        if (wholePagePartition != null) {
+            // whole input page will go to this partition, compact the input page avoid over-retaining memory and to
+            // match the behavior of sub-partitioned pages that copy positions out
+            page.compact();
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:124]: why not put this inside the `partitionPage` or inline `partitionPage` here?

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+            for (int i = 0; i < partitionSize; i++) {
+                writerAssignments[nextPartitionWriter].add(positions[i]);
+            }
+        }
+
+        // build a page for each writer
+        for (int bucket = 0; bucket < writerAssignments.length; bucket++) {
+            IntArrayList positionsList = writerAssignments[bucket];
+            int bucketSize = positionsList.size();
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:169]: nit: I would drop `bucketSize` and use `positionsList.size()` directly, the name is confusing.

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    private int getNextPartitionWriter(int partition, long[] partitionPhysicalWrittenBytes)
+    {
+        PartitionScalingInfo scalingInfo = partitionScalingInfos[partition];
+        int maxWriterCount = scalingInfo.getMaxWriterCount();
+
+        // Scale up writers for a partition when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes is greater than writerMinSize.
+        // This also mean that we won't scale local writers if the writing speed can cope up
+        // with incoming data. In another word, buffer utilization is below 50%.
+        if (maxWriterCount <= (buffers.size() - initialWriterCount) && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:200]: what does this condition mean `maxWriterCount <= (buffers.size() - initialWriterCount`?

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+            }
+        }
+
+        // Find the next writer in a round-robin fashion.
+        int nextWriterCount = scalingInfo.getNextWriterCount();
+
+        // If the next writer is equal to zero, then we will choose between initial writers using a simple mod
+        // between the partition id and initialWriterCount.
+        // If the next writer is above zero which means scaling has happened, then we will choose among the unused writers.
+        return nextWriterCount == 0 ? partition & initialWriterMask : initialWriterCount + nextWriterCount - 1;
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:214]: Do you think it's worth using a mask here instead of %?

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+            int initialWriterCount,
+            Supplier<long[]> partitionPhysicalWrittenBytesSupplier,
+            DataSize writerMinSize)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.maxBufferedBytes = maxBufferedBytes;
+        this.partitionedPagePreparer = requireNonNull(partitionedPagePreparer, "partitionedPagePreparer is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        checkArgument(Integer.bitCount(initialWriterCount) == 1, "initialWriterCount must be a power of 2");
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:95]: why do we need this constriant?

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 16]
```diff
@@ -0,0 +1,580 @@
...
+import static io.trino.sql.planner.plan.TableWriterNode.InsertTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.RefreshMaterializedViewTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.TableExecuteTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.WriterTarget;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.function.Function.identity;
+
+/**
+ * It partitions the pages and then create a separate pageSink for each partitioned page.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:76]: nit: `It partitions the pages` -> `Partitions the pages`

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java (hunk 9 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:79]: Consider adding `WriterUtils` class for shared code

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 17]
```diff
@@ -0,0 +1,580 @@
...
+            this.pageSinkManager = requireNonNull(pageSinkManager, "pageSinkManager is null");
+            requireNonNull(writerTarget, "writerTarget is null");
+            checkArgument(
+                    writerTarget instanceof CreateTarget
+                            || writerTarget instanceof InsertTarget
+                            || writerTarget instanceof RefreshMaterializedViewTarget
+                            || writerTarget instanceof TableExecuteTarget,
+                    "writerTarget must be CreateTarget, InsertTarget, RefreshMaterializedViewTarget or TableExecuteTarget");
+            this.target = writerTarget;
+            this.session = session;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:127]: it's missing rnn

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 18]
```diff
@@ -0,0 +1,580 @@
...
+
+            PartitionFunction partitionFunction = partitionFunctionFactory.create(
+                    session,
+                    partitioning,
+                    partitionChannels,
+                    partitionChannelTypes,
+                    Optional.empty(),
+                    partitionCount);
+
+            Function<Page, Page> partitionPagePreparer;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:156]: nit: why not to move that to `PartitionFunction` itself?

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 19]
```diff
@@ -0,0 +1,580 @@
...
+            if (isSystemPartitioning(partitioning)) {
+                partitionPagePreparer = identity();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:158]: add a comment when it can be a system partitioned (aka preferred insert (arbitrary) partitioning)

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 20]
```diff
@@ -0,0 +1,580 @@
...
+        timer.end(statisticsTiming);
+        ListenableFuture<Void> blockedOnAggregation = statisticAggregationOperator.isBlocked();
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedOnWrites = ImmutableList.builder();
+        blockedOnWrites.add(blockedOnAggregation);
+
+        Page partitionedPage = partitionPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:361]: `If there are more partitions than` -> `If there are more physical partitions than.., then it's possible that multiple physical partitions`

File: core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java [hunk 21]
```diff
@@ -0,0 +1,580 @@
...
+        // unique id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java:362]: `unique id` -> `bucket id`

File: core/trino-main/src/main/java/io/trino/operator/PipelineContext.java [hunk 22]
```diff
@@ -327,6 +329,25 @@ public long getPhysicalWrittenDataSize()
...
+    {
+        // Avoid using stream api for performance reasons
+        int maxSize = 0;
+        List<long[]> partitionPhysicalWrittenDataSizes = new ArrayList<>();
+        for (DriverContext context : drivers) {
+            long[] partitionPhysicalWrittenDataSize = context.getPartitionPhysicalWrittenDataSize();
+            maxSize = max(maxSize, partitionPhysicalWrittenDataSize.length);
+            partitionPhysicalWrittenDataSizes.add(partitionPhysicalWrittenDataSize);
+        }
+        long[] res = new long[maxSize];
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PipelineContext.java:342]: `res` -> `result`

File: core/trino-main/src/test/java/io/trino/operator/TestPartitionedTableWriterOperator.java [hunk 23]
```diff
@@ -0,0 +1,544 @@
...
+        @Override
+        public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle)
+        {
+            ConnectorPageSink pageSink = pageSinkSupplier.get();
+            pageSinks.add(pageSink);
+            return pageSink;
+        }
+
+        @Override
+        public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorInsertTableHandle insertTableHandle)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/TestPartitionedTableWriterOperator.java:441]: nit: consider testing with system partitioning

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPartitioningHandle.java [hunk 24]
```diff
@@ -24,4 +24,10 @@ default boolean isCoordinatorOnly()
     {
         return false;
     }
+
+    // It helps the engine to identify whether writer scaling per partition is allowed
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPartitioningHandle.java:28]: `It helps the engine to identify whether writer scaling per partition is allowed`
->
`Specifies if writing to partition has to be performed by a single writer instance`

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 25]
```diff
@@ -0,0 +1,253 @@
...
+    {
+        PartitionScalingInfo scalingInfo = partitionScalingInfos[partition];
+        int maxWriterCount = scalingInfo.getMaxWriterCount();
+
+        // Scale up writers for a partition when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes is greater than writerMinSize.
+        // This also mean that we won't scale local writers if the writing speed can cope up
+        // with incoming data. In another word, buffer utilization is below 50%.
+        if (maxWriterCount <= (buffers.size() - initialWriterCount) && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
+            long physicalWrittenBytes = partition < partitionPhysicalWrittenBytes.length ? partitionPhysicalWrittenBytes[partition] : 0;
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:201]: How this will detect that we have heavy used partition? same with `physicalWrittenBytes >= writerMinSize * maxWriterCount` after some time (and size) we will just start writers if we have memory buffers filled.

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+
+    @Override
+    public ListenableFuture<Void> waitForWriting()
+    {
+        return memoryManager.getNotFullFuture();
+    }
+
+    private static class PartitionScalingInfo
+    {
+        private int maxWriterCount = 1;
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:232]: nit: Maybe different variable name as it this points to how many writers where created and not a max we could have and both could mix together.

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+        if (maxWriterCount <= (buffers.size() - initialWriterCount) && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
+            long physicalWrittenBytes = partition < partitionPhysicalWrittenBytes.length ? partitionPhysicalWrittenBytes[partition] : 0;
+            if (physicalWrittenBytes >= writerMinSize * maxWriterCount) {
+                scalingInfo.incrementWriterCount();
+                log.debug("Increased task writer count to " + (maxWriterCount + 1) + " for partition " + partition);
+            }
+        }
+
+        // Find the next writer in a round-robin fashion.
+        int nextWriterCount = scalingInfo.getNextWriterCount();
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java:209]: is this related to count? it is more like getting next worker id/pointer

File: core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java [hunk 28]
```diff
@@ -3145,7 +3149,8 @@ public PhysicalOperation visitRefreshMaterializedView(RefreshMaterializedViewNod
         public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPlanContext context)
         {
             // Set table writer count
-            int writerCount = isLocalScaledWriterExchange(node.getSource()) ? getTaskScaleWritersMaxWriterCount(session) : getTaskWriterCount(session);
+            Optional<ExchangeNode> scaleWriterExchangeNode = getScaleWriterExchangeNode(session, node.getSource());
+            int writerCount = scaleWriterExchangeNode.isPresent() ? getTaskScaleWritersMaxWriterCount(session) : getTaskWriterCount(session);
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java:3153]: Is this general rule that we are using Optional `isPresent` with ternary instead of we use map/orElseGet?

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java [hunk 29]
```diff
@@ -0,0 +1,253 @@
...
+    @Override
+    public ListenableFuture<Void> waitForWriting()
+    {
+        return memoryManager.getNotFullFuture();
+    }
+
+    private static class PartitionScalingInfo
+    {
+        private int maxWriterCount = 1;
+        private int currentWriterCount = -1;
```
[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java]: > then adding 1 before modulo would not be required

WDYM?

File: core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java [hunk 30]
```diff
@@ -88,7 +88,7 @@ public LocalExchange(
             int defaultConcurrency,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java:83]: `Refactor LocalExchange` -> `Pass partition channel types directly to LocalExchange`

File: core/trino-main/src/main/java/io/trino/operator/AbstractTableWriterOperator.java [hunk 31]
```diff
@@ -0,0 +1,355 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.allAsList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.trino.operator.OperationTimer.OperationTiming;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public abstract class AbstractTableWriterOperator
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/AbstractTableWriterOperator.java:48]: Add commit that renames current operator first, then add a commit that makes it abstract

File: core/trino-main/src/main/java/io/trino/operator/AbstractTableWriterOperator.java [hunk 32]
```diff
@@ -0,0 +1,355 @@
...
+        blockedOnWrites.add(blockedOnAggregation);
+        blockedOnWrites.addAll(writePage(page));
+
+        updateMemoryUsage();
+        blocked = asVoid(allAsList(blockedOnWrites.build()));
+        rowCount += page.getPositionCount();
+        updateWrittenBytes();
+    }
+
+    protected abstract List<ListenableFuture<?>> writePage(Page page);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/AbstractTableWriterOperator.java:144]: group all `abstract protected` methods below public methods

File: core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+                    partitionChannelTypes,
+                    bucketToPartition);
+        }
+
+        return new BucketPartitionFunction(
+                nodePartitioningManager.getBucketFunction(session, partitioning, partitionChannelTypes, bucketCount),
+                bucketToPartition);
+    }
+
+    public Function<Page, Page> createPartitionPagePreparer(PartitioningHandle partitioning, List<Integer> partitionChannels)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PartitionFunctionFactory.java:96]:
```
public class SelectChannels
  implements PartitionFunction {
  public SelectChannels(PartitionFunction delegate, PartitioningHandle partitioning, List<Integer> partitionChannels) {
    ..
  }
}
```

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPartitioningHandle.java [hunk 34]
```diff
@@ -24,4 +24,10 @@ default boolean isCoordinatorOnly()
     {
         return false;
     }
+
+    // Specifies if writing to partition has to be performed by a single writer instance
+    default boolean isSingleWriterPerPartition()
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPartitioningHandle.java:29]: partitioning handle is insert agnostic. After some thought, it's better to put it in `ConnectorMetadata`:
```
boolean ConnectorMetadata#isSingleWriterPerPartition(ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
```

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 35]
```diff
@@ -79,6 +79,7 @@
     public static final String SCALE_WRITERS = "scale_writers";
     public static final String TASK_SCALE_WRITERS_ENABLED = "task_scale_writers_enabled";
     public static final String TASK_SCALE_WRITERS_MAX_WRITER_COUNT = "task_scale_writers_max_writer_count";
+    public static final String TASK_SCALE_WRITERS_PARTITION_COUNT = "task_scale_writers_partition_count";
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:82]: I think it's very technical. Let's just keep it as config (maybe) or just some high enough number (like 10k).

File: core/trino-main/src/main/java/io/trino/operator/DriverContext.java [hunk 36]
```diff
@@ -275,6 +277,22 @@ public long getPhysicalWrittenDataSize()
...
+            context.getPartitionPhysicalWrittenBytes().forEach((partition, physicalWrittenBytes) ->
+                    result.compute(partition, (key, value) -> {
+                        if (value == null) {
+                            return physicalWrittenBytes;
+                        }
+                        return physicalWrittenBytes + value;
+                    }));
+        }
+        return result;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/DriverContext.java:293]: return `ImmutableMap.copyOf(...)`

File: core/trino-main/src/main/java/io/trino/operator/PipelineContext.java [hunk 37]
```diff
@@ -327,6 +329,22 @@ public long getPhysicalWrittenDataSize()
...
+        for (DriverContext context : drivers) {
+            context.getPartitionPhysicalWrittenBytes().forEach((partition, physicalWrittenBytes) ->
+                    result.compute(partition, (key, value) -> {
+                        if (value == null) {
+                            return physicalWrittenBytes;
+                        }
+                        return physicalWrittenBytes + value;
+                    }));
+        }
+        return result;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PipelineContext.java:345]: return `ImmutableMap.copyOf(...)`

**Comment by gaurav8297** [comment]
Closing this in favor of https://github.com/trinodb/trino/pull/14718
